src/software/new_geom/convex_polygon.cpp

```diff
@@ -167,3 +167,68 @@ double ConvexPolygon::area() const
 
     return std::abs(first_term - second_term) / 2;
 }
+
+Point ConvexPolygon::centroid() const
+{
+    // Adapted from
+    // https://www.geeksforgeeks.org/find-the-centroid-of-a-non-self-intersecting-closed-polygon/
+    double x_centre    = 0;
+    double y_centre    = 0;
+    double signed_area = 0;
+
+    for (unsigned i = 0; i < points_.size(); i++)
+    {
+        double x0 = points_[i].x();
+        double y0 = points_[i].y();
+        double x1 = points_[(i + 1) % points_.size()].x();
+        double y1 = points_[(i + 1) % points_.size()].y();
+        double a  = (x0 * y1) - (x1 * y0);
+
+        x_centre += (x0 + x1) * a;
+        y_centre += (y0 + y1) * a;
+        signed_area += a;
+    }
+
+    return Point((Vector(x_centre, y_centre) / (3 * signed_area)));
+}
+
+ConvexPolygon ConvexPolygon::expand(const Vector& v) const
+{
+    // ASCII art showing an expanded ConvexPolygon
+    //
+    // Original ConvexPolygon:
+    //
+    //            B---------C
+    //            |         |
+    //            |         |
+    //            A---------D
+    //
+    //                 |
+    //                 V
+    //          expansion vector
+    //
+    // Expanded ConvexPolygon (A and D are shifted down):
+    //
+    //            B---------C
+    //            |         |
+    //            |         |
+    //            |         |
+    //            A'--------D'
+    //
+
+    std::vector<Point> expanded_points;
+    Point c = centroid();
+
+    for (const auto& p : points_)
+    {
+        if ((p - c).project(v).normalize() == v.normalize())
```

Comment:
Because this is geometry, I need to ask if `==` is numerically fragile here? May want to do a comparison using `GeomConstants::EPSILON`

Reply:
OK I changed it to use orientation (which side of the line perpendicular to v along c). The only fp error issue is if the point is close to the line, which shouldn't be a problem because it will decide one way or the other and "double down" on that decision (either by shifting the point even further away from the centroid, or by leaving the point as is and pulling the centroid away from the point
https://stackoverflow.com/questions/1560492/how-to-tell-whether-a-point-is-to-the-right-or-left-side-of-a-line